### ISSUE TITLE
Parallel assemble and EliminateEssentialBC

### ISSUE THREAD

**Issue by kvoronin** [mfem/mfem#181]
Hello,

I am trying to solve a least-squares problem in MFEM in Hdiv space, and got stuck with the example code. In serial (or with -n 1) it works fine but produces wrong result (error is too large and depends on the number of processes) in parallel. I think there is something wrong in the way I do the assembling and elimination of essential boundary conditions or handling parallel objects.

Maybe someone can point out what is wrong with the following lines?
```
    ParFiniteElementSpace *R_space = new ParFiniteElementSpace(pmesh.get(), hdiv_coll);
    ParGridFunction * x(new ParGridFunction(R_space));
    *x = 0.0;
    HypreParVector *X = x->ParallelAverage();

    // Setting boundary conditions.
    Array<int> ess_tdof_list, ess_bdr(pmesh->bdr_attributes.Max());
    ess_bdr = 0;
    ess_bdr[0] = 1; // t = 0
    ess_bdr[1] = 1; // lateral boundary
   
    // Define the parallel linear forms.

   ParLinearForm *fform(new ParLinearForm(R_space));
   fform->AddDomainIntegrator(new VectordivDomainLFIntegrator(some_coefficient));
   fform->Assemble();

   // Assemble the finite element matrix for the operator  A
   // A = (sigma, tau)_{L2} +  (div sigma, div tau)_{L2}
   ParBilinearForm *Ablock(new ParBilinearForm(R_space));
   Ablock->AddDomainIntegrator(new DivDivIntegrator(one));
   Ablock->AddDomainIntegrator(new VectorFEMassIntegrator(some_coefficient));
   Ablock->Assemble();

    x->ProjectCoefficient(exact_solution);
    Ablock->EliminateEssentialBC(ess_bdr,*X,*fform);
    Ablock->Finalize();

    HypreParMatrix *A;
    A = Ablock->ParallelAssemble();

    HypreParVector *B = fform->ParallelAssemble();
    *X = 0.0;
    //... then I proceed with solving linear system
```
Maybe some transition to the true dofs is missing? If there is nothing obvious, I can also cut out the MWE.

Thanks,
Kirill

**Comment by vladotomov** [comment]
You should use the parallel version of the elimination, see ParBilinearForm::ParallelEliminateEssentialBC( ).

You may also consider using ParBilinearForm::FormLinearSystem( ), it will take care of the boundary conditions.

Hope this helps, Vladimir

**Comment by kvoronin** [comment]
Thanks, @vladotomov!
Before reading your answer, I used
```
   ParFiniteElementSpace *R_space = new ParFiniteElementSpace(pmesh.get(), hdiv_coll);
   ParGridFunction * x(new ParGridFunction(R_space));
   *x = 0.0;
   HypreParVector *X = x->ParallelAverage();

    // Setting boundary conditions.
    Array<int> ess_tdof_list, ess_bdr(pmesh->bdr_attributes.Max());
    ess_bdr = 0;
    ess_bdr[0] = 1; // t = 0
    ess_bdr[1] = 1; // lateral boundary

   // 9. Define the parallel forms
   ParLinearForm *fform(new ParLinearForm(R_space));
   fform->AddDomainIntegrator(new VectordivDomainLFIntegrator(somecoeff));
   fform->Assemble();

   // 10. Assemble the operator  A
   //     A = ( sigma, tau)_{L2} + (div sigma, div tau)_{L2}

   ParBilinearForm *Ablock(new ParBilinearForm(R_space));
   Ablock->AddDomainIntegrator(new DivDivIntegrator(onecoeff));
   Ablock->AddDomainIntegrator(new VectorFEMassIntegrator(matrixcoeff));

    Ablock->Assemble();
    Ablock->EliminateEssentialBC(ess_bdr,*x,*fform);
    Ablock->Finalize();

    HypreParMatrix *A;
    A = Ablock->ParallelAssemble();
    HypreParVector *B = fform->ParallelAssemble();

    *X = 0.0;
```
which works.

Could you explain your answer a little bit more? ParallelEliminateEssentialBC has hypre parallel matrix and vectors as arguments, so how this routine should be called?
I tried, following your suggestion, to assemble like this:
```
   ...
   Ablock->AddDomainIntegrator(new VectorFEMassIntegrator(matrixcoeff));

    HypreParVector *B = fform->ParallelAssemble();
    HypreParMatrix *A;
    A = Ablock->ParallelAssemble();
    Ablock->ParallelEliminateEssentialBC(ess_bdr,*A, *X, *B);

```
but it doesn't work (segfaults).

Thanks,
Kirill

**Comment by vladotomov** [comment]
Yes, it's fine to eliminate the BC in serial, and then to form the hypre structures. I got confused, because I saw the capital X in your first post:
Ablock->EliminateEssentialBC(ess_bdr, *X, *fform);
and thought that you were working on the true dofs.

The call to ParallelEliminate looks correct, I'm not sure about the segfault. I'll have to see the full code (did you call Ablock.Assemble and Finalize).
